### ISSUE TITLE
Crash when `normal :` command executes and user is inputting command at the same time

### ISSUE THREAD

**Issue by xndcn** [vim/vim#3492]
Try this crash.vimrc
```vim
func Handler(id)
    exec "normal :echomsg 'Hello World'\n"
endfunc

call timer_start(1000, 'Handler')
```

Reproduce steps:
1. Open Vim `vim -u crash.vimrc`
2. Input `:` or `/`
3. Wait a second, and after `Hello World` message is shown, input any keys (except `Enter`...)

Vim will crash immediately.
Tested with `master` branch build, on Ubuntu 18.04

It seems that when user input `:` or `/`, `getcmdline` will be invoked and then waiting next keys.
But `exec "normal :"` will also invoke `getcmdline`, and make `ccline.cmdbuff` empty in the end.
Since `ccline` is a static global variable, any following input keys will wake up `getcmdline` then crash.

The best solution may be like the TODO saying, which seems a huge job:
```c
/* TODO: make it local to getcmdline() and pass it around. */
static struct cmdline_info ccline;
```
For workaround, save the last `ccline` when `ccline.cmdbuff` is not NULL and then restore is working. However, the message `Hello World` will be missed.
In another way, check `ccline.cmdbuff` after `c = safe_vgetc();` and overwrite the key `c = ESC` is also working.

**Comment by brammool** [comment]
> Try this crash.vimrc
> ```vim
> func Handler(id)
>     exec "normal :echomsg 'Hello World'\n"
> endfunc
> 
> call timer_start(1000, 'Handler')
> ```
> 
> Reproduce steps:
> 1. Open Vim `vim -u crash.vimrc`
> 2. Input `:` or `/`
> 3. Wait a second, and after `Hello World` message is shown, input any keys (except `Enter`...)
> 
> Vim will crash immediately.
> Tested with `master` branch build, on Ubuntu 18.04
> 
> It seems that when user input `:` or `/`, `getcmdline` will be invoked and then waiting next keys.
> But `exec "normal :"` will also invoke `getcmdline`, and make `ccline.cmdbuff` empty in the end.
> Since `ccline` is a static global variable, any following input keys will wake up `getcmdline` then crash.
> 
> The best solution may be like the TODO saying, which seems a huge job:
> ```c
> /* TODO: make it local to getcmdline() and pass it around. */
> static struct cmdline_info ccline;
> ```
> For workaround, save the last `ccline` when `ccline.cmdbuff` is not NULL and then restore is working. However, the message `Hello World` will be missed.
> In another way, check `ccline.cmdbuff` after `c = safe_vgetc();` and overwrite the key `c = ESC` is also working.

Thanks for reporting.  getcmdline() currently saves the line when doing
something that might call it recursively.  The timer can happen at any time
vgetc() is called.  Saving and restoring then is inefficient.

I think we should turn it around: When getcmdline() is called it checks if it
is a recursive call, and saves the cmdline then.  That's going to be a bit of
work, but the result should be better.  It's a bit like your patch, but
dropping the other places where ccline is saved and restored, to make it
consistent.

We don't know what the timer is going to do, it may also interact with the
current command line.  Thus making the command line private and pass it around
probably doesn't work.

-- 
hundred-and-one symptoms of being an internet addict:
117. You are more comfortable typing in html.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by xndcn** [comment]
@brammool Thanks! I just forgot the recursively invoking possibility of `getcmdline()`...
It seems a lot of work and hard for me...
Maybe `prev_ccline` could be used to check whether it is a recursive call ?

**Comment by brammool** [comment]
Fixed by Patch 8.1.0439
